### PR TITLE
[ISSUE #71] RocketMQ DataSource for Spark SQL

### PR DESCRIPTION
**Pull request by fuyufjh**
Please do not create a Pull Request without creating an issue first. 

## What is the purpose of the change

Implement the input source of RocketMQ for Spark SQL and Spark Structured Streaming ISSUE: #71 

## Brief changelog

* Bump version of Spark to 2.3.0
* Implement `RocketMQSourceProvider` with following traits (features):
  - `StreamSourceProvider`: RocketMQ as source for streaming query
  - `StreamSinkProvider`: RocketMQ as sink for streaming query
  - `RelationProvider`: RocketMQ as source for batch query
  - `CreatableRelationProvider`: RocketMQ as sink for batch query 

## Verifying this change

* Check `RocketMQDataSourceTest`

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by vongosling** [comment]
@vesense

**Comment by vongosling** [comment]
LGTM, This is an awesome pr. Consider the sparse unit-test, I will merge it and polish here. @fuyufjh Did you test it in the production environment?

**Comment by vesense** [comment]
@fuyufjh @vongosling 
I just took a look. Overall it LGTM. This PR supported Spark 2.0 new APIs. 
There are three points should be noticed: 
1. Usage documents need to be added.
2. The code refer to [spark-kafka-sql](https://github.com/apache/spark/tree/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010), some of the code are copied from that(e.g. `RocketMQOffsetRangeLimit`,`JsonUtils`,`RocketMQSourceOffset`). We should add the merit for origin author.
3. I didn't see `RocketMQSink`. Do we have a plan to add a `org.apache.spark.sql.execution.streaming.Sink`?  It is easy to do. You can refer to [KafkaSink](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSink.scala)

@fuyufjh Thanks for your great job. Keep the good work.

**Comment by vongosling** [comment]
@fuyufjh 
The code refer to spark-kafka-sql, some of the code are copied from that(e.g. RocketMQOffsetRangeLimit,JsonUtils,RocketMQSourceOffset). We should add the merit for origin author.

It would be helpful to add the copy description as @vesense said. If you could help to add it, I will merge and test again :-)

**Comment by vongosling** [comment]
@fuyufjh ping

**Comment by vesense** [comment]
>Really agree with this! I'll complete this work later (as well as other suggestions). This may take some days and I'll ping you back once done.

@fuyufjh Great! I will have a look again once the work is completed.

@vongosling About point 1 and 2, maybe I can give a help in another PR if @fuyufjh no time.
 So, I'm +1 for merging this patch.

**Comment by fuyufjh** [comment]
stylecheck failed but its not my fault :(

```
/Users/eric/Workspace/github.com/apache/rocketmq-externals/rocketmq-spark/src/main/java/org/apache/rocketmq/spark/RocketMQConfig.java:46:115: error: Only one statement per line allowed.
/Users/eric/Workspace/github.com/apache/rocketmq-externals/rocketmq-spark/src/main/java/org/apache/rocketmq/spark/streaming/MessageSet.java:31:67: error: '{' is not preceded with whitespace.
/Users/eric/Workspace/github.com/apache/rocketmq-externals/rocketmq-spark/src/main/java/org/apache/rocketmq/spark/streaming/DefaultMessageRetryManager.java:29:71: error: '{' is not preceded with whitespace.
/Users/eric/Workspace/github.com/apache/rocketmq-externals/rocketmq-spark/src/main/java/org/apache/rocketmq/spark/streaming/DefaultMessageRetryManager.java:30:24: error: ',' is not followed by whitespace.
/Users/eric/Workspace/github.com/apache/rocketmq-externals/rocketmq-spark/src/main/java/org/apache/rocketmq/spark/streaming/DefaultMessageRetryManager.java:91:37: error: ',' is not followed by whitespace.
/Users/eric/Workspace/github.com/apache/rocketmq-externals/rocketmq-spark/src/main/java/org/apache/rocketmq/spark/streaming/RocketMQReceiver.java:63:80: error: 'typecast' is not followed by whitespace.
```

**Comment by fuyufjh** [comment]
@vongosling @vesense 

Hello guys, as you seen I have made lots of changes to the PR branch, including

- [x] Support writing streaming data or result of batch query to RocketMQ
- [x] Remove a bunch of noisy redundant code, such as fixes for KAFKA-1894
- [x] Modify the user interface, such as parameters, to make it looks more user-friendly
- [x] Refine the tests
- [x] Add [documents](https://github.com/fuyufjh/rocketmq-externals/blob/feature/spark-sql-rocketmq-datasource/rocketmq-spark/spark-structured-streaming-rocketmq.md)

And to answer some of the comments above,

> Did you test it in the production environment? 

NO. Due to some reason we changed the plan so this work is/will NOT be part of a production. :( Wish someone could do more tests.

> The code refer to spark-kafka-sql, some of the code are copied from that. We should add the merit for origin author.

Yes you are right. Sadly I am not pretty sure how to do this in a proper way. Could you please give me a help?

Please inform me if I missed something else. Thanks!

**Comment by fuyufjh** [comment]
@vongosling done

**Comment by justinmclean** [comment]
As the code was under ALv2 you may need to do a little more than just credit where it came from. See for example http://www.apache.org/dev/licensing-howto.html#alv2-dep

File: rocketmq-spark/src/main/scala/org/apache/spark/sql/rocketmq/CachedRocketMQProducer.scala [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * This file was taken from Apache Spark org/apache/spark/sql/kafka010/CachedKafkaProducer.scala
+ *
```
Comment (vongosling) [inline rocketmq-spark/src/main/scala/org/apache/spark/sql/rocketmq/CachedRocketMQProducer.scala]: excellent job

**Comment by vongosling** [comment]
@fuyufjh @vesense @justinmclean I will merge it, thanks @fuyufjh.

**Pull request merged**